l1-contracts/test/foundry/unit/concrete/upgrades/BaseZkSyncUpgrade.t.sol

```diff
@@ -0,0 +1,285 @@
+    }
+
+    // Patch upgrade can't set upgrade txn
+    function test_revertWhen_PatchCantSetUpgradeTxn() public {
+        // Change bootload and default account hashes to 0, to skip previous path only checks
+        proposedUpgrade.bootloaderHash = bytes32(0);
+        proposedUpgrade.defaultAccountHash = bytes32(0);
+
+        baseZkSyncUpgrade.setProtocolVersion(SemVer.packSemVer(0, 1, 0));
+        proposedUpgrade.newProtocolVersion = SemVer.packSemVer(0, 1, 1);
+
+        vm.expectRevert(abi.encodeWithSelector(PatchCantSetUpgradeTxn.selector));
+        baseZkSyncUpgrade.upgrade(proposedUpgrade);
+    }
+
+    // L2 upgrade nonce is not equal to the new protocol version
+    function test_revertWhen_L2UpgradeNonceIsNotEqualToNewProtocolVersion(
+        uint32 newProtocolVersion,
+        uint32 nonce
+    ) public {
+        vm.assume(newProtocolVersion > 0);
+        vm.assume(nonce != newProtocolVersion && nonce > 0);
+
+        uint256 semVerNewProtocolVersion = SemVer.packSemVer(0, newProtocolVersion, 0);
+
+        baseZkSyncUpgrade.setProtocolVersion(SemVer.packSemVer(0, newProtocolVersion - 1, 0));
+
+        proposedUpgrade.newProtocolVersion = semVerNewProtocolVersion;
+        proposedUpgrade.l2ProtocolUpgradeTx.nonce = nonce;
+
+        vm.expectRevert(
+            abi.encodeWithSelector(L2UpgradeNonceNotEqualToNewProtocolVersion.selector, nonce, newProtocolVersion)
+        );
+        baseZkSyncUpgrade.upgrade(proposedUpgrade);
+    }
+
+    // Wrong number of factory deps
+    function test_revertWhen_WrongNumberOfFactoryDeps(
+        uint8 factoryDepsLength,
+        uint8 l2ProtocolUpgradeTxFDLength
+    ) public {
+        vm.assume(factoryDepsLength != l2ProtocolUpgradeTxFDLength);
+
+        proposedUpgrade.l2ProtocolUpgradeTx.factoryDeps = new uint256[](l2ProtocolUpgradeTxFDLength);
+
+        vm.expectRevert(abi.encodeWithSelector(TooManyFactoryDeps.selector));
+        baseZkSyncUpgrade.upgrade(proposedUpgrade);
+    }
+
+    // Factory deps can be at most 32
+    function test_revertWhen_FactoryDepsCanBeAtMost32(uint8 maxNewFactoryDeps) public {
+        vm.assume(maxNewFactoryDeps > MAX_NEW_FACTORY_DEPS);
+
+        proposedUpgrade.l2ProtocolUpgradeTx.factoryDeps = new uint256[](maxNewFactoryDeps);
+
+        vm.expectRevert(abi.encodeWithSelector(TooManyFactoryDeps.selector));
+        baseZkSyncUpgrade.upgrade(proposedUpgrade);
+    }
+
+    // Wrong factory dep hash
+    function test_revertWhen_WrongFactoryDepHash() public {
+        bytes[] memory factoryDeps = new bytes[](1);
+        factoryDeps[0] = "11111111111111111111111111111111";
+
+        proposedUpgrade.l2ProtocolUpgradeTx.factoryDeps = new uint256[](1);
+
+        bytes32 bytecodeHash = L2ContractHelper.hashL2Bytecode(factoryDeps[0]);
+
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                PreviousUpgradeNotFinalized.selector,
+                bytes32(proposedUpgrade.l2ProtocolUpgradeTx.factoryDeps[0])
```

Comment:
the test logic seem to be wrong, it says "wrong factory deps", but in fact the upgrade expects to receive the factory dep in the error

Reply:
Yeah, this test was targeted towards an earlier version (v-25). Changed it.